Maven CodeSnippet Plugin (#544)
Azure/azure-sdk-tools: packages/java-packages/snippet-replacer-maven-plugin/README.md

# Snippet Replacer

> `snippet-replacer-maven-plugin` allows java devs to reference actual java code in their javadoc comments. Developers running code written by the Azure team want their documentation to be useful. This plugin helps the Azure SDK team deliver on that promise by ensuring that code samples presented in doc-comments is always _actual running code_.

Reviewer: ```suggestion
`snippet-replacer-maven-plugin` allows Java developers to reference verified Java code in their Javadoc comments. Developers running code written by the Azure team want their documentation to be useful. This plugin helps the Azure SDK team deliver on that promise by ensuring that code samples presented in doc-comments are always _actual running code_.
```

> `snippet-replacer-maven-plugin` allows java devs to reference actual java code in their javadoc comments. Developers running code written by the Azure team want their documentation to be useful. This plugin helps the Azure SDK team deliver on that promise by ensuring that code samples presented in doc-comments is always _actual running code_.

Reviewer 2: This tool is very useful for any Java repo and if there's nothing in the implementation that is Azure SDK specific, we should make the documentation Azure SDK agnostic. 

So, it would be good to just replace "This plugin helps the Azure SDK team deliver on that promise" with "This plugin allows Java developers to ensure code samples presented in doc-comments is always _actual running code_".

### How does it work?

First, reference the plugin in your maven project's `pom.xml` file.

```xml
<plugin>
  <groupId>com.azure.tools</groupId>
  <artifactId>snippet-replacer-maven-plugin</artifactId>
  <version>1.0.0</version>
  <configuration>
    <mode>update</mode>
    <targetDir>${project.basedir}</targetDir>
  </configuration>
  <executions>
    <execution>
      <phase>process-sources</phase>
      <goals>
        <goal>snippet-engine</goal>
      </goals>
    </execution>
  </executions>
</plugin>

```

The plugin is intended to be run in the `process-sources` phase. 

### The modes of operation

`snippet-replacer` has two modes, `update` and `verify`.

`update` mode runs during your build and _actually updates_ your Javadoc comments with referenced source code.

> `verify` mode is intended to be run during CI/PR builds. It will error a  javadoc comment that is not updated.

Reviewer: ```suggestion
`verify` mode is intended to be run during CI/PR builds. It will error if a Javadoc comment doesn't match what is generated.
```
